### PR TITLE
feat: Qdrant ingest destination

### PR DESCRIPTION
**Pull request by Anush008**
This PR intends to add [Qdrant](https://qdrant.tech/) as a supported ingestion destination.

- Implements CLI and programmatic usage.
- Documentation update
- Integration test script

### CODE REVIEW

**Comment by Anush008** [comment]
@Coniferish, @scanny, could you please review?

File: test_unstructured_ingest/dest/qdrant.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+           -p 6333:$QDRANT_PORT \
+           --name $CONTAINTER_NAME qdrant/qdrant:latest
+
+trap stop_docker SIGINT
+trap stop_docker ERR
+
+until curl --output /dev/null --silent --get --fail http://$QDRANT_HOST/collections; do
+  printf 'Waiting for Qdrant server to start...'
+  sleep 5
+done
```
Comment (ryannikolaidis) [inline test_unstructured_ingest/dest/qdrant.sh:34]: shouldn't be an issue, but maybe timeout and error after a certain number of retries?

**Comment by ryannikolaidis** [comment]
when running the test locally I'm hitting an error for "No buffer space available". any ideas?

here's the trace:

```

--- Logging error ---
multiprocessing.pool.RemoteTraceback:
"""
Traceback (most recent call last):
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/backends/sync.py", line 38, in write
    n = self._sock.send(buffer)
OSError: [Errno 55] No buffer space available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
    yield
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_transports/default.py", line 218, in handle_request
    resp = self._pool.handle_request(req)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 253, in handle_request
    raise exc
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py", line 237, in handle_request
    response = connection.handle_request(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/connection.py", line 90, in handle_request
    return self._connection.handle_request(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 112, in handle_request
    raise exc
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 82, in handle_request
    self._send_request_body(**kwargs)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 135, in _send_request_body
    self._send_event(event, timeout=timeout)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_sync/http11.py", line 144, in _send_event
    self._network_stream.write(bytes_to_send, timeout=timeout)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/backends/sync.py", line 35, in write
    with map_exceptions(exc_map):
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc)
httpcore.WriteError: [Errno 55] No buffer space available

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api_client.py", line 101, in send_inner
    response = self._client.send(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_client.py", line 908, in send
    response = self._send_handling_auth(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_client.py", line 936, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_client.py", line 973, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_client.py", line 1009, in _send_single_request
    response = transport.handle_request(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_transports/default.py", line 217, in handle_request
    with map_httpcore_exceptions():
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.WriteError: [Errno 55] No buffer space available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/connector/qdrant.py", line 105, in upsert_batch
    response = client.upsert(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/qdrant_client.py", line 987, in upsert
    return self._client.upsert(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/qdrant_remote.py", line 1300, in upsert
    http_result = self.openapi_client.points_api.upsert_points(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api/points_api.py", line 1439, in upsert_points
    return self._build_for_upsert_points(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api/points_api.py", line 738, in _build_for_upsert_points
    return self.api_client.request(
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api_client.py", line 74, in request
    return self.send(request, type_)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api_client.py", line 91, in send
    response = self.middleware(request, self.send_inner)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api_client.py", line 200, in __call__
    return call_next(request)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/qdrant_client/http/api_client.py", line 103, in send_inner
    raise ResponseHandlingException(e)
qdrant_client.http.exceptions.ResponseHandlingException: [Errno 55] No buffer space available

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/error.py", line 19, in wrapper
    return f(*args, **kwargs)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/utils.py", line 214, in wrapper
    return func(*args, **kwargs)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/connector/qdrant.py", line 109, in upsert_batch
    raise WriteError(f"Qdrant error: {api_error}") from api_error
unstructured.ingest.error.WriteError: Qdrant error: [Errno 55] No buffer space available

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/error.py", line 22, in wrapper
    raise cls(cls.error_string.format(str(error))) from error
unstructured.ingest.error.DestinationConnectionError: Error in connecting to downstream data source: Qdrant error: [Errno 55] No buffer space available
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/cli/base/dest.py", line 70, in dest
    runner.run(**parent_options)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/runner/base_runner.py", line 41, in run
    self.process_documents(
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/runner/base_runner.py", line 80, in process_documents
    process_documents(
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/processor.py", line 90, in process_documents
    pipeline.run()
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/pipeline/pipeline.py", line 113, in run
    self.write_node(iterable=partitioned_jsons)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/pipeline/interfaces.py", line 65, in __call__
    self.result = self.run(iterable)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/pipeline/write.py", line 18, in run
    self.dest_doc_connector.write(docs=ingest_docs)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/connector/qdrant.py", line 156, in write
    self.write_dict(dict_list=dict_list)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/connector/qdrant.py", line 128, in write_dict
    pool.map(self.upsert_batch, list(chunk_generator(dict_list, qdrant_batch_size)))
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
unstructured.ingest.error.DestinationConnectionError: Error in connecting to downstream data source: Qdrant error: [Errno 55] No buffer space available

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/logger.py", line 83, in redact_jsons
    formatted_j = json.dumps(json.loads(j))
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/logging/__init__.py", line 1100, in emit
    msg = self.format(record)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/logging/__init__.py", line 943, in format
    return fmt.format(record)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/logger.py", line 94, in format
    return redact_jsons(s)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/logger.py", line 85, in redact_jsons
    formatted_j = json.dumps(ast.literal_eval(j))
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 110, in literal_eval
    return _convert(node_or_string)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 92, in _convert
    return set(map(_convert, node.elts))
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 109, in _convert
    return _convert_signed_num(node)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 83, in _convert_signed_num
    return _convert_num(node)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 74, in _convert_num
    _raise_malformed_node(node)
  File "/Users/ryannikolaidis/.pyenv/versions/3.10.12/lib/python3.10/ast.py", line 71, in _raise_malformed_node
    raise ValueError(msg + f': {node!r}')
ValueError: malformed node or string on line 1: <ast.Name object at 0x2bf089600>
Call stack:
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/./unstructured/ingest/main.py", line 11, in <module>
    main()
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/./unstructured/ingest/main.py", line 7, in main
    ingest_cmd()
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/ryannikolaidis/.pyenv/versions/unstructured/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/ryannikolaidis/Development/unstructured/unstructured/unstructured/ingest/cli/base/dest.py", line 72, in dest
    logger.error(e, exc_info=True)
Message: DestinationConnectionError('Error in connecting to downstream data source: Qdrant error: [Errno 55] No buffer space available')
Arguments: ()
Error: Error in connecting to downstream data source: Qdrant error: [Errno 55] No buffer space available
++ stop_docker
++ docker stop qdrant_test
qdrant_test
```

**Review by ryannikolaidis — COMMENTED** [review]
@Anush008 just a quick question about retry in the test and an issue I hit running the test locally. otherwise the changes look great! thanks for contributing here and excited to get this one in!

File: test_unstructured_ingest/dest/qdrant.sh (hunk 1 above)
Comment (Anush008) [reply test_unstructured_ingest/dest/qdrant.sh]: 5ee1b7f49ea9336b72a313071d5e76d189b8b05d

**Comment by Anush008** [comment]
Hey @ryannikolaidis. The retries limit should be resolved now. 5ee1b7f49ea9336b72a313071d5e76d189b8b05d.
I am unable to reproduce the test error you are running into, did you change any config values or limits when running?

File: unstructured/ingest/cli/cmds/qdrant.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                help="The location of the Qdrant cluster.",
+            ),
+            click.Option(
+                ["--url"],
+                type=str,
+                help="The location of the Qdrant cluster.",
+            ),
+            click.Option(
+                ["--port"],
+                type=int,
```
Comment (rbiseck3) [inline unstructured/ingest/cli/cmds/qdrant.py:35]: For anything that has a default, can we also set the default parameter in the click option? 
```suggestion
                type=int,
                default=6333
```

File: unstructured/ingest/cli/cmds/qdrant.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                help="Port of the gRPC interface. Default: 6334.",
+            ),
+            click.Option(
+                ["--prefer-grpc"],
+                type=bool,
+                help="Whether to use gPRC interface whenever possible in methods. Default: False.",
+            ),
+            click.Option(
+                ["--https"],
+                type=bool,
```
Comment (rbiseck3) [inline unstructured/ingest/cli/cmds/qdrant.py:50]: for boolean types, click also supports a `is_flag` parameter which sets it just by the presence of the cli option.  
```suggestion
                type=bool,
                is_flag=True
```

**Comment by ryannikolaidis** [comment]
> Hey @ryannikolaidis. The retries limit should be resolved now. [5ee1b7f](https://github.com/Unstructured-IO/unstructured/commit/5ee1b7f49ea9336b72a313071d5e76d189b8b05d). I am unable to reproduce the test error you are running into, did you change any config values or limits when running?

@Anush008 I hadn't changed anything, though just gave my machine a reboot and everything worked perfectly, so chalking this up to something ephemeral to the state of my machine. maybe docker weirdness.

**Comment by ryannikolaidis** [comment]
@Anush008 code looks good. final steps: can you bump the version and changelog. also looks like the shell lint checks are failing. probably running `make tidy-shell` should clean that up. (may need a `make tidy` too, once changelog is bumped we'll see how the linting checks perform)

**Comment by Anush008** [comment]
Hey @ryannikolaidis. Could you rerun the workflows please?

**Comment by Anush008** [comment]
The failed tests seems to be arising from missing credentials of other workflows.

**Comment by ryannikolaidis** [comment]
@Anush008 this all looks great, thanks for the contribution here! 

closing this and opened #2338 so that we can run it completely through CI and merge.

**Review by ryannikolaidis — APPROVED** [review]

**Pull request closed without merging**